### ISSUE TITLE
As a developer, I want to spend 10% less time debugging and testing cellxgene features. (a.k.a. cellxgene test improvement plan)

### ISSUE THREAD

**Issue by maniarathi** [chanzuckerberg/single-cell#44]
**Sub-story**: As a developer, I need my test results to provide meaningful feedback on whether my code has broken anything or caused any incompatibilities. For example, breaking the compatibility tests should give me information on how I might have made cellxgene incompatible with our most important dependency, `anndata`, which would cause pain later on to our users.
**In plain English**: Fix the compatibility tests in cellxgene.
**Need**: Developers need reliable, running tests that tell us if we've added code that will cause dependency failures for desktop cellxgene users.

**Sub-story**: As a developer, I need to spend less time debugging to figure out why my change might have caused a deep-dependency change. I want confidence that each layer's interfaces have been thoroughly tested, especially around the edge-cases, so that if I create a new module that depends on it, I have confidence that I have not introduced a dependency bug that is hard to debug and takes up a lot of time. These errors also pass down to our desktop users who report back bugs that generally get added directly to our Sprint pipeline and cause expensive context shifts during a Sprint, reducing our velocity.
**In plain English**: Improve unit test coverage in cellxgene from 64% to 90%.
**Need**: Developers need thorough unit tests so that behavior of modules is explainable, predictable, and (more) bug-free. This will also inform us when we have introduced code that might have broken cellxgene in certain circumstances that would later be potentially discovered by the scientist using the desktop version.

**Sub-story**: As a developer, I need to know when I've introduced code that significantly reduced the performance of cellxgene. A "selling" point of cellxgene is that scientists can interact with their datasets at interactive speed. As a developer, I need to know when I have detracted from that experience so that I don't spend time later down the line debugging the performance issues which are sometimes tricky to find.
**In plain English**: Introduce a performance test suite to cellxgene.
**Need**: As a scientist, I want to continue having interactive speed of cellxgene even as my datasets get larger to up to a few million cells.

**Sub-story**: As a developer, I would like to avoid spending 30 minutes to 1 hour every week manually testing the cellxgene staging environment prior to release. Manual testing is error prone ([this bug](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/cellxgene/1808) wasn't caught for nearly 1 month) and time consuming. As a developer, I'd like to have confidence in my smoke tests to capture user interactions more thoroughly and consistently so that we have fewer regressions and few aborted deployments.
**In plain English**: Add the red colored use cases in the [QA script](https://docs.google.com/document/d/1nHdd8cDlmauv27oEemlMy_mEa0Dw7UMCp-w50IhNuK0/edit) to the cellxgene smoketest suite.
**Need**: ~8 engineering hours per week are spent towards smoke-testing cellxgene deployments. This time could be better spent elsewhere while an automated smoketest runs to provide more consistent results on whether the latest cellxgene deployment has any user experience regressions.

**Comment by brianraymor** [comment]
Per September 1 Refinement, @maniarathi to refine into multiple issues or epics which are more _bite-size_.

**Comment by tihuan** [comment]
QA script: https://docs.google.com/document/d/1nHdd8cDlmauv27oEemlMy_mEa0Dw7UMCp-w50IhNuK0/edit#heading=h.ytxuwawsxq0z

**Comment by maniarathi** [comment]
Closing this since only one epic remains.
